cabal-install/src/Distribution/Client/FetchUtils.hs

```diff
@@ -118,6 +122,30 @@ checkRepoTarballFetched repo pkgid = do
       then return (Just file)
       else return Nothing
 
+verifyFetchedTarball :: Verbosity -> RepoContext -> Repo -> PackageId -> IO Bool
+verifyFetchedTarball verbosity repoCtxt repo pkgid =
+   let file = packageFile repo pkgid
+       handleError :: IO Bool -> IO Bool
+       handleError act = do
+         res <- Safe.try act
+         case res of
+           Left e -> warn verbosity ("Error verifying fetched tarball: " ++ show (e :: SomeException)) >> pure False
+           Right b -> pure b
+   in handleError $ case repo of
+           -- a secure repo has hashes we can compare against to confirm this is the correct file.
+           RepoSecure{} ->
+             repoContextWithSecureRepo repoCtxt repo $ \repoSecure ->
+                  Sec.withIndex repoSecure $ \callbacks -> (do
+                    fileInfo <- Sec.indexLookupFileInfo callbacks pkgid
+                    infoVerified <-  do
+                      sz <- Sec.FileLength . fromInteger <$> getFileSize file
+                      if sz /= Sec.fileInfoLength (Sec.trusted fileInfo)
+                      then return False
+                      else Sec.compareTrustedFileInfo (Sec.trusted fileInfo) <$> Sec.computeFileInfo (Sec.Path file :: Sec.Path Sec.Absolute)
+                    if infoVerified then return True else warn verbosity "whoops" >>  return False) -- TODO
```

Comment:
When you get around to this TODO, could you also make this more readable? The `do` in parens is hard to parse. E.g., 

```hs
Sec.withIndex repoSecure $ \callbacks -> do
  let verification = do
        fileInfo <- Sec.indexLookupFileInfo callbacks pkgid
        sz <- Sec.FileLength . fromInteger <$> getFileSize file
        if sz /= Sec.fileInfoLength (Sec.trusted fileInfo)
        then return False
        else Sec.compareTrustedFileInfo (Sec.trusted fileInfo) <$> Sec.computeFileInfo (Sec.Path file :: Sec.Path Sec.Absolute)
  infoVerified <- 
    verification
    `Sec.catchChecked` (\(e :: Sec.InvalidPackageException) -> warn verbosity (show e) >> return False)
    `Sec.catchChecked` (\(e :: Sec.VerificationError) -> warn verbosity (show e) >> return False)
  if infoVerified then return True else warn verbosity "whoops" >>  return False) -- TODO
```

which also changes the semantics slightly, probably resulting in a less puzzling warning if an exception occurs.

Reply:
Thanks for noticing the TODO. I refactored and cleaned up, but could not factor out into a let block as suggested -- the checked exceptions mechanism in hackage-security is fiddly, and trying some obvious things still resulted in a situation where the typechecker yelled at me. I'm sure it could be made to work, but I think the code after the refactor is sufficiently more clear.